test/functional/services/find.js

```diff
@@ -105,143 +81,156 @@ export function FindProvider({ getService }) {
+      log.debug(`Wait for element become visible: ${selector} with timeout=${timeout}`);
+      await driver.wait(until.elementIsVisible(element._webElement), timeout);
+      return wrap(element);
     }
 
-    async findDisplayedByLinkText(selector, timeout = defaultFindTimeout) {
-      log.debug('Find.byLinkText: ' + selector);
-      return await this._ensureElementWithTimeout(timeout, async leadfoot => {
-        return wrap(await leadfoot.findDisplayedByLinkText(selector));
-      });
+    async byLinkText(selector, timeout = defaultFindTimeout) {
+      log.debug(`Find.byLinkText('${selector}') with timeout=${timeout}`);
+      return wrap(await driver.wait(until.elementLocated(By.linkText(selector)), timeout));
     }
 
     async byPartialLinkText(partialLinkText, timeout = defaultFindTimeout) {
-      log.debug(`find.byPartialLinkText(${partialLinkText})`);
-      return await this._ensureElementWithTimeout(timeout, async leadfoot => {
-        return wrap(await leadfoot.findByPartialLinkText(partialLinkText));
-      });
+      log.debug(`Find.byPartialLinkText('${partialLinkText}')  with timeout=${timeout}`);
+      return wrap(await driver.wait(until.elementLocated(By.partialLinkText(partialLinkText)), timeout));
     }
 
     async exists(findFunction, timeout = WAIT_FOR_EXISTS_TIME) {
-      return await this._withTimeout(timeout, async leadfoot => {
-        try {
-          await findFunction(leadfoot);
-          return true;
-        } catch (error) {
-          return false;
+      await this._withTimeout(timeout);
+      try {
+        const found = await findFunction(driver);
+        await this._withTimeout(defaultFindTimeout);
+        if (Array.isArray(found)) {
+          return found.length > 0;
+        } else {
+          return found instanceof WebElementWrapper;
         }
-      });
+      } catch (err) {
+        await this._withTimeout(defaultFindTimeout);
+        return false;
+      }
     }
 
     async existsByLinkText(linkText, timeout = WAIT_FOR_EXISTS_TIME) {
-      log.debug(`existsByLinkText ${linkText}`);
-      return await this.exists(async leadfoot => wrap(await leadfoot.findDisplayedByLinkText(linkText)), timeout);
+      log.debug(`Find.existsByLinkText('${linkText}')  with timeout=${timeout}`);
+      return await this.exists(async driver => wrapAll(await driver.findElements(By.linkText(linkText))), timeout);
     }
 
     async existsByDisplayedByCssSelector(selector, timeout = WAIT_FOR_EXISTS_TIME) {
-      log.debug(`existsByDisplayedByCssSelector ${selector}`);
-      return await this.exists(async leadfoot => wrap(await leadfoot.findDisplayedByCssSelector(selector)), timeout);
+      log.debug(`Find.existsByDisplayedByCssSelector('${selector}') with timeout=${timeout}`);
+      return await this.exists(async (driver) => {
+        const elements = wrapAll(await driver.findElements(By.css(selector)));
+        return await this.filterElementIsDisplayed(elements);
```

Comment:
Filtering after the `findElements()` call like this means that the timeout isn't used if the `findElements()` call returns elements but none of them are displayed, I'm _pretty sure_ this is the reason we're seeing flakiness like https://kibana-ci.elastic.co/job/elastic+kibana+master/66/JOB=x-pack-ciGroup6,node=immutable/testReport/junit/X-Pack%20Chromium%20Functional%20Reporting%20Tests/x-pack_test_reporting_functional_reporting%C2%B7js/reporting_app_Reporting_Dashboard_Print_PDF_button_becomes_available_when_saved/